src/components/CodeEditor.tsx

```diff
@@ -331,6 +337,43 @@ export const CodeEditor = ({
       }
     }
   }, [files[currentFile], currentFile])
+
+  useEffect(() => {
+    const handleBeforeUnload = (event: BeforeUnloadEvent) => {
+      if (hasUnsavedChanges) {
+        event.preventDefault()
+        event.returnValue = "" // Shows the confirmation dialog on reload or close if there are unsaved changes
+      }
+    }
+
+    const handlePopState = (event: PopStateEvent) => {
+      if (hasUnsavedChanges) {
+        const userConfirmed = window.confirm(
+          "Are you sure you want to go back? Unsaved changes may be lost.",
+        )
+        if (!userConfirmed) {
+          // Push the state again to stay on the same page
+          window.history.pushState(null, "", window.location.href)
+        }
+      }
+    }
+
+    // Attach event listeners
+    window.addEventListener("beforeunload", handleBeforeUnload)
+    window.addEventListener("popstate", handlePopState)
+
+    // Push the initial state only once, when component mounts
+    if (hasUnsavedChanges) {
+      window.history.pushState(null, "", window.location.href)
+    }
+
+    // Cleanup event listeners on component unmount
+    return () => {
+      window.removeEventListener("beforeunload", handleBeforeUnload)
+      window.removeEventListener("popstate", handlePopState)
+    }
+  }, [hasUnsavedChanges])
```

Comment:
this component (the `CodeEditor`) is already really big, can you put this code in the parent component? Also can you make a dedicated hook e.g. `useWarnUserForUnsavedChanges(hasUnsavedChanges)`

Reply:
i created  a hook for the functionality now we just have to call it. it is in hooks folder with name useWarnUserForUnsavedChanges. I also reverted this code to previous state i.e.
const match = line.match(/@tsci\/[\w\-.]+/)

I hope there is no more conflicts and hope it helps.